pypesto/select/method.py

```diff
@@ -0,0 +1,512 @@
+
+
+class MethodSignalProceed(str, Enum):
+    """Indicators for how a model selection method should proceed."""
+
+    # TODO move to PEtab Select?
+    STOP = 'stop'
+    CONTINUE = 'continue'
+
+
+@dataclass
+class MethodSignal:
+    """The state of a model selection method after a single model calibration.
+
+    Attributes
+    ----------
+    accept:
+        Whether to accept the model.
+    proceed:
+        How the method should proceed.
+    """
+
+    accept: bool
+    # TODO change to bool?
+    proceed: MethodSignalProceed
+
+
+class MethodLogger:
+    """Log results from a model selection method.
+
+    Attributes
+    ----------
+    level:
+        The logging level.
+    logger:
+        A logger from the `logging` module.
+    """
+
+    def __init__(
+        self,
+        level: str = 'info'
+    ):
+        self.logger = logging.getLogger(__name__)
+        self.level = level
+
+    def log(self, message, level: str = None) -> None:
+        """Log a message.
+
+        Parameters
+        ----------
+        message:
+            The message.
+        level:
+            The logging level. Defaults to the value defined in the
+            constructor.
+        """
+        if level is None:
+            level = self.level
+        getattr(self.logger, level)(message)
+
+    def new_selection(self) -> None:
+        """Start logging a new model selection."""
+        padding = 20
+        self.log('-'*padding + 'New Selection' + '-'*padding)
+        columns = [
+            "Predecessor model subspace:ID",
+            "Model subspace:ID",
+            "Criterion ID",
+            "Predecessor model criterion",
+            "Model criterion",
+            "Criterion difference",
+            "Accept",
+        ]
+        self.log("\t".join([f'[{c}]' for c in columns]))
+
+    def new_result(
+        self,
+        accept,
+        criterion,
+        model,
+        predecessor_model,
+        max_id_length: str = 12,
+        precision: int = 3,
+    ) -> None:
+        """Log a model calibration result.
+
+        Parameters
+        ----------
+        accept:
+            Whether the model is accepted.
+        criterion:
+            The criterion type.
+        max_id_length:
+            Model and predecessor model IDs are truncated to this length in the
+            logged message.
+        model:
+            The calibrated model.
+        predecessor_model:
+            The predecessor model.
+        precision:
+            The number of decimal places to log.
+        """
+        model_criterion = model.get_criterion(criterion)
+
+        def get_model_id(model: Model) -> str:
+            """Get a model ID for logging.
+
+            Parameters
+            ----------
+            model:
+                The model.
+
+            Returns
+            -------
+            str
+                The ID.
+            """
+            model_subspace_id = model.model_subspace_id or ''
```

Comment:
this syntax is not very common, as it can be misinterpreted. more common: `A = A if A else B`. but whatever